src/script/cpp_api/s_async.cpp

```diff
@@ -101,38 +111,41 @@ void AsyncEngine::addWorkerThread()
 }
 
 /******************************************************************************/
-u32 AsyncEngine::queueAsyncJob(std::string &&func, std::string &&params,
-		const std::string &mod_origin)
+
+u32 AsyncEngine::queueAsyncJob(LuaJobInfo &&job)
 {
 	MutexAutoLock autolock(jobQueueMutex);
 	u32 jobId = jobIdCounter++;
 
-	jobQueue.emplace_back();
-	auto &to_add = jobQueue.back();
-	to_add.id = jobId;
-	to_add.function = std::move(func);
-	to_add.params = std::move(params);
-	to_add.mod_origin = mod_origin;
+	job.id = jobId;
+	jobQueue.push_back(std::move(job));
 
 	jobQueueCounter.post();
 	return jobId;
 }
 
-u32 AsyncEngine::queueAsyncJob(std::string &&func, PackedValue *params,
+u32 AsyncEngine::queueAsyncJob(std::string &&func, std::string &&params,
 		const std::string &mod_origin)
 {
-	MutexAutoLock autolock(jobQueueMutex);
-	u32 jobId = jobIdCounter++;
+	LuaJobInfo to_add(std::move(func), std::move(params), mod_origin);
+	return queueAsyncJob(std::move(to_add));
+}
 
-	jobQueue.emplace_back();
-	auto &to_add = jobQueue.back();
-	to_add.id = jobId;
-	to_add.function = std::move(func);
-	to_add.params_ext.reset(params);
-	to_add.mod_origin = mod_origin;
+u32 AsyncEngine::queueAsyncJob(std::string &&func, PackedValue *params,
+		const std::string &mod_origin)
+{
+	LuaJobInfo to_add(std::move(func), params, mod_origin);
+	return queueAsyncJob(std::move(to_add));
+}
 
-	jobQueueCounter.post();
-	return jobId;
+bool AsyncEngine::cancelAsyncJob(const u32 &id)
+{
+	MutexAutoLock autolock(jobQueueMutex);
+	int pos = id - (jobIdCounter - jobQueue.size());
```

Comment:
this probably goes wrong if the counter ever overflows (so that `id > jobIdCounter`)
it will definitely go wrong on overflow if `jobIdCounter < jobQueue.size()`

Reply:
Ok, Seems like iterating the job queue is the way to go then.

(That said, I personally consider an ID overflow to be unlikely, although it is better to stay safe.)